File: server/services/shared_config.py
"""
Shared configuration and models for AI Saathi agent services
"""
import google.generativeai as genai
from google.generativeai import types as genai_types
import json
import os
from typing import Dict, List, Any, Optional
from pydantic import BaseModel
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Configure Gemini API
gemini_api_key = os.getenv('GEMINI_API_KEY')
logger.info(f"Gemini API Key: {gemini_api_key[:8] + '...' if gemini_api_key else 'Not found'}")

is_valid_api_key = gemini_api_key and len(gemini_api_key) > 20
logger.info(f"Is valid API key: {is_valid_api_key}")

# Configure Gemini
if is_valid_api_key:
    genai.configure(api_key=gemini_api_key)
    # Use Gemini 1.5 Pro for enhanced performance
    model = genai.GenerativeModel('gemini-1.5-pro-latest')
else:
    model = None

# Request Models
class TeachingAidRequest(BaseModel):
    subject: str
    grade: int
    topic: str
    language: str
    materials: List[str]
    type: str  # "worksheet" | "flashcard" | "story"

class LessonPlanRequest(BaseModel):
    subject: str
    grades: List[int]
    timeLimit: int
    topic: str
    language: str
    materials: List[str]

class AssessmentRequest(BaseModel):
    subject: str
    grade: int
    topic: str
    language: str
    questionCount: int

class TranslationRequest(BaseModel):
    text: str
    fromLanguage: str
    toLanguage: str

class StoryRequest(BaseModel):
    theme: str
    grades: List[int]
    language: str
    moral: str
    characters: List[str]

def parse_gemini_response(content: str) -> Dict[str, Any]:
    """Parse Gemini response, handling markdown formatting"""
    content = content.strip()
    if content.startswith('```json'):
        content = content.replace('```json', '').replace('```', '').strip()
    elif content.startswith('```'):
        content = content[3:-3].strip()
    
    return json.loads(content)

File: server/services/shared-config.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { z } from 'zod';

// Configure logging
const logger = {
  info: (message: string) => console.log(`[INFO] ${message}`),
  error: (message: string) => console.error(`[ERROR] ${message}`),
  warn: (message: string) => console.warn(`[WARN] ${message}`)
};

// Configure Gemini API
const geminiApiKey = process.env.GEMINI_API_KEY;
logger.info(`Gemini API Key: ${geminiApiKey ? geminiApiKey.substring(0, 8) + '...' : 'Not found'}`);

const isValidApiKey = geminiApiKey && geminiApiKey.length > 20;
logger.info(`Is valid API key: ${isValidApiKey}`);

// Configure Gemini
let model: any = null;
if (isValidApiKey) {
  const genAI = new GoogleGenerativeAI(geminiApiKey);
  // Use Gemini 1.5 Pro for enhanced performance
  model = genAI.getGenerativeModel({ model: 'gemini-1.5-pro-latest' });
}

// Request Models using Zod for validation
export const TeachingAidRequestSchema = z.object({
  subject: z.string(),
  grade: z.number(),
  topic: z.string(),
  language: z.string(),
  materials: z.array(z.string()),
  type: z.enum(['worksheet', 'flashcard', 'story'])
});

export const LessonPlanRequestSchema = z.object({
  subject: z.string(),
  grades: z.array(z.number()),
  timeLimit: z.number(),
  topic: z.string(),
  language: z.string(),
  materials: z.array(z.string())
});

export const AssessmentRequestSchema = z.object({
  subject: z.string(),
  grade: z.number(),
  topic: z.string(),
  language: z.string(),
  questionCount: z.number()
});

export const TranslationRequestSchema = z.object({
  text: z.string(),
  fromLanguage: z.string(),
  toLanguage: z.string()
});

export const StoryRequestSchema = z.object({
  theme: z.string(),
  grades: z.array(z.number()),
  language: z.string(),
  moral: z.string(),
  characters: z.array(z.string())
});

export type TeachingAidRequest = z.infer<typeof TeachingAidRequestSchema>;
export type LessonPlanRequest = z.infer<typeof LessonPlanRequestSchema>;
export type AssessmentRequest = z.infer<typeof AssessmentRequestSchema>;
export type TranslationRequest = z.infer<typeof TranslationRequestSchema>;
export type StoryRequest = z.infer<typeof StoryRequestSchema>;

export function parseGeminiResponse(content: string): any {
  /**Parse Gemini response, handling markdown formatting*/
  content = content.trim();
  if (content.startsWith('```json')) {
    content = content.replace('```json', '').replace('```', '').trim();
  } else if (content.startsWith('```')) {
    content = content.substring(3, content.length - 3).trim();
  }
  
  try {
    return JSON.parse(content);
  } catch (error) {
    logger.error(`Failed to parse Gemini response: ${error}`);
    return { error: 'Failed to parse response', rawContent: content };
  }
}

export { model, isValidApiKey, logger }; 